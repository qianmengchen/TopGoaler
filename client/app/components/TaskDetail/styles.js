import { StyleSheet } from 'react-native';

const list = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'stretch'
  }
});

const frequency = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 10
  },
  textgroup: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  }
});

const status = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5,
    marginTop: 10,
    marginHorizontal: 80
  },
  accept: {
    backgroundColor: 'blue'
  },
  done: {
    backgroundColor: 'red',
    paddingHorizontal: 15
  },
  unfollow: {
    backgroundColor: 'yellowgreen',
    paddingHorizontal: 15
  },
  multipleBtns: {
    flexDirection: 'row',
    justifyContent: 'space-evenly'
  },
  buttonText: {
    color: '#fff',
    textAlign: 'center',
    fontSize: 17,
    paddingVertical: 10
  }
});

export { list, frequency, status };
