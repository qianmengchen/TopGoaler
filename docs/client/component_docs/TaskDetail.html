<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: TaskDetail</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: TaskDetail</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>TaskDetail<span class="signature">()</span><span class="type-signature"></span></h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    

    
    <h4 class="name" id="TaskDetail"><span class="type-signature"></span>new TaskDetail<span class="signature">()</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    This is the component that shows all the information of a task in that channel.
It also allows the user to enroll/drop that task.
</div>













    <h5 class="subsection-title">Properties:</h5>

    

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>title</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The title of the task.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>period</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A type encoding indicating the day, week, or month period of a task.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>pattern</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A value indicating the numbere of times a user can complete a task within a given period.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>points</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A value indicating the points associated with the task.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>enrolled</code></td>
            

            <td class="type">
            
                
<span class="param-type">boolean</span>


            
            </td>

            

            

            <td class="description last">A boolean value indicating whether the user is enrolled in a task or not.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>task_id</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A unique number identifying the task.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>user_id</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last">A unique number identifying the user.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>enroll</code></td>
            

            <td class="type">
            
                
<span class="param-type">function</span>


            
            </td>

            

            

            <td class="description last">A function allowing the user to subscribe to a task.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>drop</code></td>
            

            <td class="type">
            
                
<span class="param-type">function</span>


            
            </td>

            

            

            <td class="description last">A function allowing the user to unsubcribe from a task.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>newActivity</code></td>
            

            <td class="type">
            
                
<span class="param-type">function</span>


            
            </td>

            

            

            <td class="description last">A function allowing a user to create a new task.</td>
        </tr>

    
    </tbody>
</table>




<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="TaskDetail_TaskDetail.js.html">TaskDetail/TaskDetail.js</a>, <a href="TaskDetail_TaskDetail.js.html#line7">line 7</a>
    </li></ul></dd>
    

    

    

    
</dl>


















    
    </div>

    

    

    

    

    

    

    

    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="App.html">App</a></li><li><a href="AuthNav.html">AuthNav</a></li><li><a href="ChannelIcon.html">ChannelIcon</a></li><li><a href="ChannelIconList.html">ChannelIconList</a></li><li><a href="ChannelListPage.html">ChannelListPage</a></li><li><a href="ChannelMemberView.html">ChannelMemberView</a></li><li><a href="ChannelOverview.html">ChannelOverview</a></li><li><a href="ChannelPublicView.html">ChannelPublicView</a></li><li><a href="Feed.html">Feed</a></li><li><a href="MemberList.html">MemberList</a></li><li><a href="NewChannelPage.html">NewChannelPage</a></li><li><a href="NewTaskPage.html">NewTaskPage</a></li><li><a href="PerformancePage.html">PerformancePage</a></li><li><a href="Profile.html">Profile</a></li><li><a href="Proposal.html">Proposal</a></li><li><a href="ProposalsPage.html">ProposalsPage</a></li><li><a href="SignIn.html">SignIn</a></li><li><a href="SignUp.html">SignUp</a></li><li><a href="TaskChecklist.html">TaskChecklist</a></li><li><a href="TaskDetail.html">TaskDetail</a></li><li><a href="TaskEntry.html">TaskEntry</a></li><li><a href="TaskListPage.html">TaskListPage</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Mon Dec 03 2018 02:45:35 GMT-0800 (Pacific Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>