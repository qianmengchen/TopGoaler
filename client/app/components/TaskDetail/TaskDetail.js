import React, { Component } from 'react';
import { Text, View, TouchableHighlight } from 'react-native';
import { Card } from 'react-native-elements';
import { frequency, status } from './styles';
import { Event } from '../../constants';

class TaskDetail extends Component {
  _accept(task_id, user_id) {
    this.props.enroll(user_id, task_id);
    this.props.newActivity(user_id, task_id, Event.JOIN);
  }

  _unfollow(task_id, user_id) {
    this.props.drop(user_id, task_id);
    this.props.newActivity(user_id, task_id, Event.DROP);
  }

  render() {
    let {
      title,
      period,
      pattern,
      points,
      enrolled,
      task_id,
      user_id
    } = this.props;
    let button;

    if (!enrolled) {
      button = (
        <TouchableHighlight
          style={[status.container, status.accept]}
          underlayColor="#aaa"
          onPress={() => this._accept(task_id, user_id)}
        >
          <Text style={status.buttonText}>Accept Challenge!</Text>
        </TouchableHighlight>
      );
    } else {
      button = (
        <View style={status.multipleBtns}>
          <TouchableHighlight
            style={[status.container, status.unfollow]}
            underlayColor="#aaa"
            onPress={() => this._unfollow(task_id, user_id)}
          >
            <Text style={status.buttonText}>Unfollow</Text>
          </TouchableHighlight>
        </View>
      );
    }

    return (
      <Card title={title}>
        <View style={frequency.container}>
          <Text>
            Frequency: {pattern} time(s) per {period}
          </Text>
          <Text>Reward: {points} point(s)!</Text>
        </View>
        {button}
      </Card>
    );
  }
}

export default TaskDetail;
