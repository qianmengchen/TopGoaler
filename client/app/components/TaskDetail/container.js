import { connect } from 'react-redux';
import TaskDetail from './TaskDetail';
import { enrollTaskAsUser, newActivityLog } from '../../actions';

const mapDispatchToProps = dispatch => ({
  enroll: (user_id, task_id) => dispatch(enrollTaskAsUser(task_id, user_id)),
  newActivity: (user_id, task_id, event) =>
    dispatch(newActivityLog(task_id, user_id, event))
});

const mapStateToProps = state => ({
  user_id: state.auth.id
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TaskDetail);
