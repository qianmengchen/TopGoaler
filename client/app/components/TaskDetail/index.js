import TaskDetail from './TaskDetail';

export { TaskDetail };
