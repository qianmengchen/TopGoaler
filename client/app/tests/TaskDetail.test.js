import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunkMiddleware from 'redux-thunk';
import sinon from 'sinon';
require('isomorphic-fetch');

import TaskDetail from '../components/TaskDetail/TaskDetail';

const middlewares = [thunkMiddleware]; // you can mock any middlewares here if necessary
const mockStore = configureStore(middlewares);

const initialTaskState = {
  tasks: ['do frontend', 'do backend', 'do MVP ftw'],
  isFetching: false,
  currentTaskPage: 'ChannelListPage'
};

describe('Testing Sign In Page', () => {
  const accept = sinon.spy(TaskDetail.prototype, '_accept');
  const unfollow = sinon.spy(TaskDetail.prototype, '_unfollow');
  const done = sinon.spy(TaskDetail.prototype, '_done');

  const wrapper = shallow(<TaskDetail />, {
    context: { store: mockStore(initialTaskState) }
  });
  const render = wrapper.dive();

  it('renders as expected', () => {
    expect(render).toMatchSnapshot();
  });

  it('renders when state is set to not available', () => {
    wrapper.setState({ status: 'in-progress' });
    expect(wrapper.state('status')).toBe('in-progress');

    wrapper.setState({ status: 'done' });
    expect(wrapper.state('status')).toBe('done');

    wrapper.setState({ status: 'available' });
    expect(wrapper.state('status')).toBe('available');
  });

  it('should update status state when buttons are pressed', () => {
    wrapper.find('TouchableHighlight').forEach(child => {
      child.simulate('press');
    });

    wrapper.setState({ status: 'in-progress' });
    wrapper.find('TouchableHighlight').forEach(child => {
      child.simulate('press');
    });

    expect(accept.calledOnce).toBe(true);
    expect(done.calledOnce).toBe(true);
    expect(unfollow.calledOnce).toBe(true);
  });
});
