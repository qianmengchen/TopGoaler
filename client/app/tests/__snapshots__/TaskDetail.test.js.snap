// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Task Detail Page renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Themed.Card>
    <Component
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Frequency: 
         time(s) per 
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Reward: 
         point(s)!
      </Text>
    </Component>
    <TouchableHighlight
      activeOpacity={0.85}
      delayPressOut={100}
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 5,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginHorizontal": 80,
            "marginTop": 10,
          },
          Object {
            "backgroundColor": "blue",
          },
        ]
      }
      underlayColor="#aaa"
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 17,
            "paddingVertical": 10,
            "textAlign": "center",
          }
        }
      >
        Accept Challenge!
      </Text>
    </TouchableHighlight>
  </Themed.Card>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "Consumer": [Circular],
      "Provider": Object {
        "$$typeof": Symbol(react.provider),
        "_context": [Circular],
      },
      "_calculateChangedBits": null,
      "_changedBits": 0,
      "_currentRenderer": null,
      "_currentValue": undefined,
      "_defaultValue": undefined,
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_currentRenderer": null,
        "_currentValue": undefined,
        "_defaultValue": undefined,
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
